### ISSUE TITLE
Implementation of systhreads with pthreads

### ISSUE THREAD

**Issue by ctk21** [ocamllabs/ocaml-multicore#357]
Following further discussion around implementing systhread compatibility with [domains](https://github.com/ocaml-multicore/ocaml-multicore/pull/342#issuecomment-643119638), we should explore porting the existing systhread implementation using pthreads to multicore. 

The idea would be to implement systhread as it is done [upstream](https://github.com/ocaml/ocaml) by utilizing the `caml_enter_blocking_section`/`caml_leave_blocking_section` hooks to provide a mechanism to context switch as in `caml_thread_save_runtime_state`/`caml_thread_restore_runtime_state`. It would need to make sure that we interface with the multicore runtime to handle local roots registration, local root scanning and backup threads (#356) correctly. It should inherit the existing systhread scheduling mechanism. 

The aim is to provide an implementation which is compatible with the existing systhread. 

It is left as future work to support a more advanced threading implementation that integrates with effects, as discussed on #100.

**Comment by jhwoodyatt** [comment]
Presumably this would still provide the global thread interlock behavior, i.e. only one thread is allowed to proceed at any particular time even where threads are spawned from multiple domains. Is that right?

**Comment by gasche** [comment]
Naive question: I assumed that Threads would be implemented with fibers on a single domain. Is that not a sensible step forward anymore? My understanding of the state of fibers was that they are not efficient enough to be exposed to users right now (I remember a Discuss post mentioning that their implementation would need to change first to use assembly for some operations, which multicore-ocaml has decided to delay for later); but if there is an internal implementation (that is a bit slow), wouldn't that be a reasonable basis for Threads?

**Comment by ctk21** [comment]
There are currently two implemented parallelism & concurrency abstractions in multicore:
- domains: these execute in parallel and are fairly heavyweight coming with their own minor heap and the synchronization required to manage a multicore minor and major GC.
- fibers exposed through effects: fibers are light weight execution contexts with dynamic stacks that allow for flexible concurrency. fibers are (currently) created and scheduled using effects. The continuations can be passed between domains allowing for complex schedulers to be constructed on top of the multicore runtime. However fibers currently do not have any mechanism for handling blocking sections that could occur in external system or C-library calls; if a fiber enters a blocking C call, then all fibers in that domain will be blocked until the call returns.

This is where implementing systhreads with pthreads sharing a domain comes in:
- each pthread sharing a domain (or domain-execution-context / "dec"): is a concurrency abstraction that allows for blocking system and C-library calls. They are less heavyweight than domains because only one dec (that is pthread sharing a domain) can be executing OCaml code in a domain at a given time. This sharing of the domain and scheduling between decs is done at the C level using the `caml_enter_blocking_section` / `caml_leave_blocking_section` calls. 

There are two reasons for implementing this abstraction: provide a compatibility route for existing OCaml code that use systhreads (some usages have 100-1000s of concurrent pthreads, for example the [XAPI](https://github.com/ocaml-multicore/ocaml-multicore/pull/342#issuecomment-643335056)); provide a multicore abstraction to allow user level management of C-stack resources when you are forced to by external libraries or system interfaces.

**Comment by lpw25** [comment]
> provide a multicore abstraction to allow user level management of C-stack resources when you are forced to by external libraries or system interfaces.

I think it is worth pointing out that there are various ideas for how we might provide for this second use case without needing a separate threads abstraction. In particular, I think @stedolan has some plans for providing support for this without explicit threads.

So maybe we will be able to eliminate the threads abstraction in the long run, but for now the backwards compatibility reasons alone are a compelling reason to include it.

**Comment by gasche** [comment]
I hadn't thought about the interaction with blocking system calls, thanks, that clarifies it. It sounds like it could be a problem in the long run: I never thought of cooperative (instead of preemptive) concurrency for fibers as an expressiveness limitation, but not being able to use blocking syscalls in fiber-based concurrent models do sound like a limitation that it would be good to address in the future. The newly-proposed work on pthreads could actually help with that, providing an interface similar to `Lwt_detach` or Async's `In_thread` mechanism that a fiber scheduler could use.

**Comment by jhwoodyatt** [comment]
You'll still need some kind of global interlock to implement the semantics of the legacy `Thread` library, which only permits one `Thread.t` to be executing OCaml code at any one time. Accordingly, it isn't enough to limit the `pthread_t` to its associated domain— it must be that only one `pthread_t` may run _in any domain_ at any time.

**Comment by kayceesrk** [comment]
@jhwoodyatt We're not going for this semantics. Our aim is to support the current use of `Thread` library using decs. Multiple decs share a domain and only one dec will run at a time. This covers backwards compatibility to the existing `Thread` module. 

If multiple domains were created, each with their own decs, then the decs on different domains will run in parallel. The decs executing across multiple domains may be synchronized using `Mutex` and `Condition`.

**Comment by abbysmal** [comment]
Just an update about the current state of this implementation.
We now have a working implementation for domain execution contexts, that can compile and execute dune.
It passes as well the `lib-threads` testsuite, although the `systhreads` testsuite still needs some care. (forced preemption through a ticking thread triggering preemption through signals is not yet implemented, for instance.)
I will come up with a PR in the next few days and will try to outline the missing parts, as well as the next discussion points.

**Comment by jhwoodyatt** [comment]
> We're not going for this semantics.

Okay then! I'm fully in support of this decision. I think it's the right compromise. You get the global `Thread` interlock when you have only one `Domain`, as in the current OCaml runtime. And you get as many threads as you can eat and an interlock on each `Domain` when there is more than one of them. I like it.

**Comment by ctk21** [comment]
#381 is merged and so we have an implementation of systhreads with pthreads which closes out this issue.
